numpy/core/src/umath/ufunc_object.c

```diff
@@ -1748,10 +1763,148 @@ make_arr_prep_args(npy_intp nin, PyObject *args, PyObject *kwds)
+        if (list_size != nin || _has_output_coredims(ufunc)) {
+            PyErr_Format(PyExc_ValueError,
+                         "axes should be a list with an entry for all "
+                         "%d inputs and outputs; entries for outputs can only "
+                         "be omitted if none of them has core axes.",
+                         nop);
+            return -1;
+        }
+        for (i = nin; i < nop; i++) {
+            remap_axis[i] = NULL;
+        }
+    }
+    for (i = 0; i < list_size; ++i) {
+        int j, iter_axis;
+        int op_ndim, core_ndim;
+        int requested_axis = 0;
+        int have_seen_axis[NPY_MAXDIMS] = {0};
+
+        core_ndim = ufunc->core_num_dims[i];
+        if (op[i] != NULL) {
+            op_ndim = PyArray_NDIM(op[i]);
+        }
+        else {
+            op_ndim = broadcast_ndim + core_ndim;
+        }
+        item = PyList_GetItem(axes, i);
+        /* if not a tuple already, make it one; content is checked later */
+        if (!PyTuple_Check(item)) {
+            item = PyTuple_Pack(1, item);
+            if (item == NULL) {
+                return -1;
+            }
+        }
+        else {
+            Py_INCREF(item);
+        }
+        if (PyTuple_Size(item) != core_ndim) {
+            if (core_ndim == 1) {
+                PyErr_Format(PyExc_TypeError,
+                             "axes item %d should be a tuple with a "
+                             "single element, or an integer", i);
+            }
+            else {
+                PyErr_Format(PyExc_TypeError,
+                             "axes item %d should be a tuple with %d "
+                             "elements", i, core_ndim);
+            }
+            goto fail;
+        }
+        /*
+         * Now create the remap, starting with the core dimensions, and
+         * then adding the remaining axes that are to be iterated over.
+         */
+        for (j = 0; j < core_ndim; j++) {
+            requested_axis = (int)PyNumber_AsSsize_t(
+                PyTuple_GetItem(item, j), NULL);
+            if (error_converting(requested_axis) ||
+                (check_and_adjust_axis(&requested_axis, op_ndim) < 0)) {
+                goto fail;
+            }
+            if (have_seen_axis[requested_axis]) {
+                PyErr_Format(PyExc_ValueError,
+                             "axes item %d has value %d repeated",
+                             i, requested_axis);
+                goto fail;
+            }
+            have_seen_axis[requested_axis] = 1;
+            remap_axis[i][op_ndim - core_ndim + j] = requested_axis;
+        }
+        Py_DECREF(item);
+        /*
+         * Fill out the ndim-core_ndim axes that have not yet been set,
+         * using have_seen_axis to skip over the core_ndim entries set above.
+         */
+        for (j = 0, iter_axis = 0; j < op_ndim - core_ndim; j++) {
+            while (have_seen_axis[iter_axis]) {
+                iter_axis++;
```

Comment:
I'm pretty concerned that this is a buffer overflow, and the outer loop should `break` if `iter_axis > op_ndim`.

Perhaps then this loop is inside-out, and should actually be
```
for (iter_axis = 0, j = 0; iter_axis < op_ndim; iter_axis++) {
    if (have_seen_axis[iter_axis]) continue;
     remap_axis[i][j++] = iter_axis;
}
```

Reply:
I still think this loop cannot fail: there will only `core_ndim` entries in `have_seen_axis` that are set. I can use your suggested loop, but it does needless iterations.

Reply:
What if `core_ndim == NPY_MAXDIMS == op_ndim`, ie `sum(have_seen_axis) == len(have_seen_axis)`?

Reply:
Oh I see - the outer loop then doesn't run at all, which avoids the problem.

Reply:
OK, I pushed up a new version, mostly with changes in variable names, which I think make the flow considerably easier to follow.